### PR TITLE
[REF] mail: introduce relational fields on discuss model

### PR DESCRIPTION
**Pull request by alexkuhn**
Before these commits, discuss relations were difficult to maintain: relations are named by local ids and ids of record, and we must juggle all the time between these local ids and the records. Also when the record is deleted, we must add lots of code to properly delete all relations on this record. Some dedicated services that manage relations on models had complicated data-structures, such as `messageIdsByChannelId` in `message_pin` service for something equivalent to `channel.pinnedMessages`.

This commit improves the readability and maintainability of relations in discuss models:

Can define a "one" relation on a discuss model:

```js
class Message extends Record {
    author = Record.one("Persona");
}
```

Relational fields can be used to uniquely identify records:

```js
class ChatWindow extends Record {
    static id = "thread";
    thread = Record.one("Thread");
}
```

There's also support for many relations:

```js
class Thread extends Record {
    messages = Record.many("Messages");
}
// ...
thread.messages = messages;
thread.messages.push(message);
thread.messages[0];
thread.messages.filter((msg) => !msg.isEmpty);
```

And support for patching models to add relations:
```js
patch(Thread.prototype, {
    setup() {
        this.pinnedMessages = Record.many("Messages");
    },
});
```

Typing of models and their patches is managed in a `@types/models.d.ts` file:
```ts
declare module "models" {
    export interface Thread {
        pinnedMessages: Message[],
    }
}
```

When a record is deleted, it is automatically removed to all its inverse relations:
```js
thread.messages = [message];
message.delete();
thread.messages; // []
```

Introducing relational fields is a key step to reduce complexity of discuss code in the models, making code less prone to bugs, and simplify development of new features that require using a lot of relations on models.

https://github.com/odoo/enterprise/pull/47745

### CODE REVIEW

File: addons/mail/static/src/core/common/store_service.js [hunk 1]
```diff
@@ -187,9 +204,126 @@ export const storeService = {
...
+                                if (r.notIn(oldRecords)) {
+                                    oldRecords.add(r);
+                                }
+                            }
+                            for (const r of oldRecords) {
+                                this.delete(r);
+                            }
+                            for (const r of newRecords) {
+                                this.add(r);
+                            }
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/store_service.js:322]: looks wrong

File: addons/mail/static/src/core/common/store_service.js [hunk 2]
```diff
@@ -187,9 +204,126 @@ export const storeService = {
...
+                            l1.__list__ = [];
+                            for (const r2 of oldRecords) {
+                                r2.__invs__.delete(r1.localId, name);
+                            }
+                            for (const r3 of collection) {
+                                l1.__list__.push(r3.localId);
+                                r3.__invs__.add(r1.localId, name);
+                            }
+                        },
+                    });
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/store_service.js:294]: This looks also wrong (there's no `.value0`

File: addons/mail/static/src/core/common/record.js [hunk 3]
```diff
@@ -79,13 +414,51 @@ export class Record {
...
+     *
+     * @type {Map<string, string|RecordList|RecordSet>}
+     */
+    __rels__ = new Map();
+    /**
+     * Track inverse relations of current record. key is inverse field name @see Record.__invs__
+     * and value contains local ids of records that have relational value on current record on
+     * inverse field name. Note that you can look up `Record.__invs__` for inverse of inverse
+     * field name.
+     */
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/record.js:460]: Comment is incorrect. There's no longer `Record.__invs__` (it's a residue from an old implem)

File: addons/mail/tooling/types/models.d.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    export interface Message_Patch {
+        pinnedAt: string,
+    }
+
+    export interface Thread_Patch {
+        pinnedMessages: RecordSet<Models["Message"]>,
+        pinnedMessagesState: "loaded"|"loading"|"error"|undefined,
+    }
```
Comment (seb-odoo) [inline addons/mail/tooling/types/models.d.ts:40]: Can this be moved to pin folder?
It looks useless as a patch if it's defined on the main file.

File: addons/mail/static/src/core/common/record.js [hunk 5]
```diff
@@ -61,6 +61,10 @@ export class RecordInverses {
     }
 }
 
+/**
+ * @typedef {Object<number, T>} RecordListIndexes<T>
+ */
+
 /**
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/record.js:68]: Unused => remove

**Comment by seb-odoo** [comment]
@robodoo r+ delegate+ rebase-ff

**Comment by seb-odoo** [comment]
@robodoo r+

**Pull request closed without merging**